builder.go

```diff
@@ -253,24 +254,46 @@ func (k *kResolver) makeAddresses(e EndpointSlice) ([]resolver.Address, string)
 	return newAddrs, ""
 }
 
-func (k *kResolver) handle(e EndpointSlice) {
-	addrs, _ := k.makeAddresses(e)
-	if len(addrs) > 0 {
+func (k *kResolver) handle(e EndpointSlice, eventType EventType) {
+	switch eventType {
+	case Added:
+		k.currentState[e.Name] = e
+	case Modified:
+		k.currentState[e.Name] = e
+	case Deleted:
+		delete(k.currentState, e.Name)
+	default:
+
+	}
+	var currentState map[resolver.Address]bool
+	var keys []resolver.Address
+	for _, value := range k.currentState {
+		addresses, _ := k.makeAddresses(value)
+		for _, address := range addresses {
+			currentState[address] = true
+		}
+
+		keys = make([]resolver.Address, 0, len(currentState))
+		for key := range currentState {
+			keys = append(keys, key)
+		}
+	}
+	if len(keys) > 0 {
```

Comment:
Shouldn't we first build all available addresses and then deduplicate them? Your logic recreates a `keys` slice over and over again for each `k.currentState` we have (for each endpoint slice).
```suggestion
	var dedupedAddressesSet map[resolver.Address]struct{}
	for _, endpointSlice := range k.currentState {
		addresses, _ := k.makeAddresses(endpointSlice)
		for _, address := range addresses {
			dedupedAddressesSet[address] = struct{}{}
		}
	}
	
	var addresses []resolver.Address
	for address := range dedupedAddressesSet {
		addresses = append(addresses, address)
	}
	if len(addresses) > 0 {
```